File: 2.string/8.StringToInteger(atoi)/main.go
package main

// https://leetcode.com/problems/string-to-integer-atoi/

const MAX_INT = (1 << 31) - 1
const MIN_INT = -(1 << 31)

// each step: result = result * 10 + (s[i] - '0')
// make sure before handle result, we need to check result <= (MAX_INT - (s[i] - '0')) / 10
func myAtoi(s string) int {
	digitStartIndex := 0
	sign := 1
	for digitStartIndex < len(s) {
		if s[digitStartIndex] == ' ' {
			digitStartIndex++
			continue
		}
		if s[digitStartIndex] == '-' {
			sign *= -1
			digitStartIndex++
			break
		}

		if s[digitStartIndex] == '+' {
			digitStartIndex++
			break
		}
		break
	}

	result := 0
	for i := digitStartIndex; i < len(s); i++ {
		if s[i] < '0' || s[i] > '9' {
			break
		}

		if result > (MAX_INT-int(s[i]-'0'))/10 {
			if sign == -1 {
				return MIN_INT
			}
			return MAX_INT
		}

		result = result*10 + int(s[i]-'0')
	}

	return result * sign
}

File: 2.string/8.StringToInteger(atoi)/main.py
# https://leetcode.com/problems/string-to-integer-atoi/

# TC: O(n), SC: O(1)
class Solution:
	def myAtoi(self, s: str) -> int:
		startIdx = 0
		signed = 0
		for c in s:
			if c == ' ':
				startIdx += 1
				continue
			if c == '+' or c == '-':
				signed = 1 if c == '+' else -1
				startIdx += 1
				break
			break
		res = 0
		MAX_INT = 2**31 - 1
		MIN_INT = -2**31
		while startIdx < len(s):
			if not s[startIdx].isnumeric():
				break
			num = ord(s[startIdx]) - ord('0')
			# -2^31 <= res * 10 + num <= 2^31 - 1
			# (-2^31 - num) / 10 <= res <= (2^31 - 1 - num) / 10
			# if res > (MAX_INT - num) / 10:
			#     if signed == -1:
			#         res = MAX_INT + 1
			#     else:
			#         res = MAX_INT
			#	  break 
			# res = res * 10 + num
			if signed == -1:
				if -res < int((MIN_INT + num) / 10):
					return MIN_INT
			else:
				if res > int((MAX_INT - num) / 10):
					return MAX_INT
			res = res * 10 + num
			startIdx += 1
		
		res = res * signed if signed != 0 else res
		return res
		
		